Paws4Me/Paws4Me/Display And Save Pets/View Models/petLocaldatabaseViewModel.swift

```diff
@@ -0,0 +1,114 @@
+    private var petLocalDatabaseRepository: SavedPetDataRepository?
+    private weak var delegate: PetLocalDatabaseViewModelDelegate?
+    private var pets: [Pet]? = []
+    private var namePet = ""
+    private var imagePet = ""
+    var isDeleteSucess = false
+
+    var nameSinglePet: String? {
+        return namePet
+    }
+
+    var imageSinglePet: String? {
+       return imagePet
+    }
+
+    // MARK: - Constructor
+    init(repository: SavedPetDataRepository,
+         delegate: PetLocalDatabaseViewModelDelegate) {
+        self.petLocalDatabaseRepository = repository
+        self.delegate = delegate
+    }
+
+    // MARK: - Functions
+    func set(petName: String) {
+        namePet = petName
+    }
+
+    func set(imagePet: String) {
+        self.imagePet = imagePet
+    }
+
+    // MARK: - Tableview data
+    var petSavedCount: Int {
+        return pets?.count ?? 0
+    }
+
+    var isPetSucessDeleted: Bool {
+        return isDeleteSucess
+    }
+
+    func savedPet(at index: Int) -> Pet? {
+        pets?[index]
+    }
+
+    func petNameSaved(pet: Pet) -> String {
+        return pet.value(forKeyPath: "petName") as? String ?? ""
+    }
+
+    func petImageSaved(pet: Pet) -> String {
+        return pet.value(forKeyPath: "petImage") as? String ?? ""
+    }
+
+    // MARK: - Local Database function fetch
+    func fetchPetDataResults() {
+        petLocalDatabaseRepository?.fetchSavedPets { [weak self] savedPets in
+            switch savedPets {
+            case .success(let savedPetData):
+                self?.pets = savedPetData
+                self?.delegate?.reloadView()
+            case .failure:
+                self?.delegate?.showError(errorTitle: "Unable to retreive all your saved pets",
+                                          errorMessage: "There was a problem retrieving your pets",
+                                          action: .retrievedPetsSavedError)
+            }
+        }
+    }
+
+    // MARK: - Local Database function save
+    func savePetInLocalDatabase(name: String, image: String) {
+        petLocalDatabaseRepository?.savePets(namePet: name, imagePet: image ) { [weak self] savedPets in
+            switch savedPets {
+            case .success:
+                self?.delegate?.reloadView()
+            case .failure:
+                self?.delegate?.showError(errorTitle: "Unable to save \(name)",
+                                          errorMessage: "There was a problem savings",
+                                          action: .savePetsError)
+            }
+            self?.fetchPetDataResults()
+        }
+    }
+
+    // MARK: - Local Database function delete
+    func deletePetLocaldatabase(petToRemove: Pet) {
+        petLocalDatabaseRepository?.deleteSavedPet(petToRemove: petToRemove) { [weak self] savedPets in
+            switch savedPets {
+            case .success:
+                self?.isDeleteSucess = true
+                _ = self?.delegate?.isPetDeletedSuccessfully()
```

Comment:
So this still isn't quite right 😄 give me a call if you want to chat about this.

Reply:
I have implemented what I think is right way of delegating. If it is wrong I will definitely take up your offer for a call to chat about this.